### DISCUSSION TITLE
iOS app complains "an error has occured" when token expires

### DISCUSSION THREAD

**Discussion by tinythomasffm** [dani-garcia/vaultwarden#5262] [Q&A]
### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.32.5
* Web-vault version: v2024.6.2c
* OS/Arch: linux/aarch64
* Running within a container: true (Base: Alpine)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.47.0
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** DOMAIN

```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_max_note_size": 10000,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*******************",
  "domain_origin": "*****://*******************",
  "domain_path": "",
  "domain_set": true,
  "duo_context_purge_schedule": "30 * * * * *",
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "duo_use_iframe": false,
  "email_2fa_auto_fallback": false,
  "email_2fa_enforce_on_verified_invite": false,
  "email_attempts_limit": 3,
  "email_change_allowed": true,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "enable_websocket": true,
  "enforce_single_org_with_reset_pw_policy": false,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "experimental_client_feature_flags": "fido2-vault-credentials",
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": "***",
  "http_request_block_non_global_ips": true,
  "http_request_block_regex": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "increase_note_size_limit": false,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "XXX Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": false,
  "push_identity_uri": "https://identity.bitwarden.com",
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "***************",
  "smtp_from_name": "XXX Vaultwarden",
  "smtp_host": "***************************************",
  "smtp_password": null,
  "smtp_port": 25,
  "smtp_security": "off",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": null,
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "user_send_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

### Vaultwarden Build Version

v1.32.5

### Deployment method

Official Container Image

### Custom deployment method

official container image running on kubernetes (k3s)

### Reverse Proxy

ingress-nginx, deployed via helm chart ingress-nginx-4.11.3 (latest)

### Host/Server Operating System

Linux

### Operating System Version

iOS

### Clients

iOS

### Client Version

2024.11.0 (1680)

### Steps To Reproduce

1. log in iOS app
2. wait until token expires on server
3. open/unlock iOS app

### Expected Result

the sync should happen without any issues.

### Actual Result

when trying to sync, iOS app throws "an error has occured". Logging off in the app and re-login fixes the problem (until the next token expire).

### Logs

```text
[2024-12-08 11:58:10.616][request][INFO] GET /api/sync
[2024-12-08 11:58:10.617][vaultwarden::auth][ERROR] Token has expired
[2024-12-08 11:58:10.617][auth][ERROR] Unauthorized Error: Invalid claim
[2024-12-08 11:58:10.617][vaultwarden::api::core::ciphers::_][WARN] Request guard `Headers` failed: "Invalid claim".
[2024-12-08 11:58:10.617][response][INFO] (sync) GET /api/sync?<data..> => 401 Unauthorized
```

**Comment by tinythomasffm** [comment]
forgot to mention: macos browser plugins, web vault work without issues.

**Reply by timbo100** [reply]
I have this same issue.  I've followed the instructions and have done a docker compose update to the latest after doing a backup.  but my IOS app still has same error.

**Comment by BlackDex** [comment]
Do you have a HA setup?
Where is the rsa key stored?

**Comment by BlackDex** [comment]
I can't reproduce this at all.
I used an iOS device with the exact same version, changed the expire/lifetime to 10 minutes to make my life easier. 
But, i do get the same error message, but it will still be logged in and syncing after getting a new refresh token.

Ill leave it for longer now and see what happens.
But i suspect maybe your reverse proxy is  modifying responses to error pages like 401 etc... and only passes on 3xx and 2xx unmodified.

**Reply by timbo100** [reply]
I am using a reverse proxy (nginx).  should there be any changes for this configuration since vaultwarden has been updated?  If so please advise.

**Comment by tinythomasffm** [comment]
no HA setup. Nothing configured about rsa key storage, so /data/rsa_key.pem

**Comment by BlackDex** [comment]
And `/data` is a persistent volume?

**Comment by tinythomasffm** [comment]
> I can't reproduce this at all. I used an iOS device with the exact same version, changed the expire/lifetime to 10 minutes to make my life easier. But, i do get the same error message, but it will still be logged in and syncing after getting a new refresh token.
> 
> Ill leave it for longer now and see what happens. But i suspect maybe your reverse proxy is modifying responses to error pages like 401 etc... and only passes on 3xx and 2xx unmodified.

that could be. But I only noticed this now, I'm using this setup (with older version of vaultwarden) for a long time and it just came up these days. Really strange.

**Comment by tinythomasffm** [comment]
> And `/data` is a persistent volume?

yes.

**Comment by tinythomasffm** [comment]
vaultwarden is deployed as a statefulset, the volume is persistent across container restarts

**Comment by BlackDex** [comment]
Maybe your reverse proxy setup has been updated too?
Or K8s with some changes which could cause this.

**Comment by tinythomasffm** [comment]
> Maybe your reverse proxy setup has been updated too? Or K8s with some changes which could cause this.

I'm just checking that. The nginx helm chart I am using was updated to that version on Oct 8 - I cannot totally exclude that the issue came with that update, checking changelogs of that now.

**Comment by tinythomasffm** [comment]
But wouldn't that issue then hit the browser plugins as well?
They work without problems.

**Comment by tinythomasffm** [comment]
Hmm.. the nginx setup does indeed use custom error pages for 401,403,404,500,501,502,503.
I will remove 401 from that list and see if it fixes the issue. But I still wonder why this came up now - these error redirects where in place all the time and I never noticed anything. 

Will check that out.

**Comment by BlackDex** [comment]
Well, i tested it just now after waiting a very long time, and having the token only valid for 10 minutes, i was still able to sync without any issue at all.

I do not know if something else may have changed on the reserve proxy. Maybe there were some bugs or fixes which caused it not to happen before.

I'm just not able to reproduce it my self.

**Comment by tinythomasffm** [comment]
I disabled the 401 custom error message and it seems to work now. I have no idea how it could have worked before, as these were in place the whole time… I‘ll watch it a little longer, but I feel this is not an issue with vaultwarden.Thanks for looking into this!Am 08.12.2024 um 15:05 schrieb Mathijs van Veluw ***@***.***>:﻿
Well, i tested it just now after waiting a very long time, and having the token only valid for 10 minutes, i was still able to sync without any issue at all.
I do not know if something else may have changed on the reserve proxy. Maybe there were some bugs or fixes which caused it not to happen before.
I'm just not able to reproduce it my self.

**Comment by devops-nomad** [comment]
Same issue here; however, [this](https://github.com/dani-garcia/vaultwarden/discussions/5012#discussioncomment-11420825) one didn't fix the problem . I tried versions 1.32.1, 1.32.5, and 1.32.7. Unfortunately, there are no logs on the Docker instance that indicate any error. I can log in to the website and the Bitwarden desktop app but not iOS Mobile. Any thoughts are much appreciated.

**iOS Version:** 18.3
**Bitwarden App Version:** 2024.12.0
**Server Version:** 1.32.5

I'm not sure if this log is related but this is the only thing I see on docker

```
[2024-12-26 17:22:41.853][request][INFO] POST /identity/connect/token
[2024-12-26 17:22:41.856][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
[2024-12-26 17:26:41.794][request][INFO] POST /identity/connect/token
[2024-12-26 17:26:41.794][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
[2024-12-26 17:30:41.764][request][INFO] POST /identity/connect/token
[2024-12-26 17:30:41.765][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
```

**Reply by twiesing** [reply]
How did you fix it? Same here. Re-installation the app doesn't work for me, too.

**Comment by twiesing** [comment]
Having the same issue. Can not get it working with deleting the app, deleting known devices, ...

Recent Versions of iOS, Bitwarden App, Vaultwarden. Using Caddy proxy with cloudflare proxy enabled.

**Reply by twiesing** [reply]
```
app-1  | [2024-12-31 12:51:16.733][request][INFO] GET /api/accounts/revision-date
app-1  | [2024-12-31 12:51:16.733][vaultwarden::auth][ERROR] Token has expired
app-1  | [2024-12-31 12:51:16.733][auth][ERROR] Unauthorized Error: Invalid claim
app-1  | [2024-12-31 12:51:16.733][vaultwarden::api::core::accounts::_][WARN] Request guard `Headers` failed: "Invalid claim".
app-1  | [2024-12-31 12:51:16.733][response][INFO] (revision_date) GET /api/accounts/revision-date => 401 Unauthorized
app-1  | [2024-12-31 12:51:16.803][request][INFO] POST /identity/connect/token
app-1  | [2024-12-31 12:51:16.821][response][INFO] (login) POST /identity/connect/token => 200 OK
app-1  | [2024-12-31 12:51:17.419][request][INFO] GET /api/accounts/revision-date
app-1  | [2024-12-31 12:51:17.422][response][INFO] (revision_date) GET /api/accounts/revision-date => 200 OK
app-1  | [2024-12-31 12:51:17.465][request][INFO] GET /api/sync
app-1  | [2024-12-31 12:51:17.597][response][INFO] (sync) GET /api/sync?<data..> => 200 OK
```

**Reply by twiesing** [reply]
After deleting the iOS app, the error in console changes, but the "An error occured" in the app is always there:

```
app-1  | [2024-12-31 13:22:28.085][request][INFO] GET /api/sync
app-1  | [2024-12-31 13:22:28.218][response][INFO] (sync) GET /api/sync?<data..> => 200 OK
app-1  | [2024-12-31 13:22:28.225][response][INFO] (sync) GET /api/sync?<data..> => 200 OK
```

**Reply by BlackDex** [reply]
Try the testing version of the container. It might solve your issue

**Reply by twiesing** [reply]
Does not work :/

```
app-1  | [2024-12-31 13:27:04.672][request][INFO] POST /identity/accounts/prelogin
app-1  | [2024-12-31 13:27:04.674][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
app-1  | [2024-12-31 13:27:04.842][request][INFO] POST /identity/connect/token
app-1  | [2024-12-31 13:27:05.292][error][ERROR] 2FA token not provided
app-1  | [2024-12-31 13:27:05.292][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
app-1  | [2024-12-31 13:27:13.341][request][INFO] POST /identity/connect/token
app-1  | [2024-12-31 13:27:13.762][vaultwarden::api::identity][INFO] User bla@gmail.com logged in successfully. IP: XXXXXXX
app-1  | [2024-12-31 13:27:13.762][response][INFO] (login) POST /identity/connect/token => 200 OK
app-1  | [2024-12-31 13:27:13.912][request][INFO] GET /api/config
app-1  | [2024-12-31 13:27:13.912][response][INFO] (config) GET /api/config => 200 OK
app-1  | [2024-12-31 13:27:14.301][request][INFO] GET /api/sync
app-1  | [2024-12-31 13:27:14.417][response][INFO] (sync) GET /api/sync?<data..> => 200 OK
app-1  | [2024-12-31 13:27:14.861][request][INFO] GET /api/sync
app-1  | [2024-12-31 13:27:14.978][response][INFO] (sync) GET /api/sync?<data..> => 200 OK
```

Can I help you with some debug output, ...?

<img src="https://github.com/user-attachments/assets/174502e3-8944-4f29-b64f-4a3c0abcb179" width="200">

**Reply by twiesing** [reply]
The debug log level does not provide any further information. **What is important**, however, is that if I click away the error message, I have an empty vault in my iOS app, but I can create new entries. I can then also view and edit these, etc. and the ‘new’ entries are even synchronised to the other clients. So it seems that the initial sync of the existing vault is not working.

And I think it's something user-specific. Other users in my instance have no problems with the iOS apps. However, if I use my iPad instead of my iPhone, for example, I get the same error including an empty vault.

**Reply by BlackDex** [reply]
Please verify and ensure you are using either testing or latest via by checking `/admin/diagnostics`

**Reply by twiesing** [reply]
<img width="1263" alt="Bildschirmfoto 2024-12-31 um 15 56 36" src="https://github.com/user-attachments/assets/2cacde11-fa92-4979-85bc-5c3b440c3159" />

**Reply by BlackDex** [reply]
Not sure how to get debug/error messages on an iPhone, so not sure how to help there.

On Android with adb it could help with an error message.
But it must be something in the synced data.

**Reply by twiesing** [reply]
> Not sure how to get debug/error messages on an iPhone, so not sure how to help there.
> 
> 
> 
> On Android with adb it could help with an error message.
> 
> But it must be something in the synced data.

I have an android device where I could test/debug. I'll try to reproduce it there and will contact you here for debug help :-)

**Reply by BlackDex** [reply]
Of the user is part of an org, try to remove that user from those orgs and see if that solves it. You could try to change the password and rotate the key, the password can stay the same though, but the rotate key is important here. That might solve it, but that would mean something is still send wrong during sync.

**Reply by twiesing** [reply]
Password change didn't work to solve the problem. Remove the user from orgs... Mhh.. I'm the owner 🙈

**Reply by BlackDex** [reply]
And you did checked the rotate checkbox right?

Hmm only way to solve that invite a new user as owner.

**Reply by twiesing** [reply]
When I click the checkbox:

```
app-1  | [2025-01-01 16:38:51.943][request][INFO] POST /api/accounts/key
app-1  | [2025-01-01 16:38:52.035][vaultwarden::api::core::accounts][ERROR] All existing ciphers must be included in the rotation
app-1  | [2025-01-01 16:38:52.037][response][INFO] (post_rotatekey) POST /api/accounts/key => 400 Bad Request
```

**Reply by BlackDex** [reply]
Do you have ssh keys?

**Reply by twiesing** [reply]
I activated the feature a few days ago `EXPERIMENTAL_CLIENT_FEATURE_FLAGS=ssh-key-vault-item,ssh-agent,` because I wanted to test the feature, but the import-button didn't worked in the app/web (both latest versions), so I disabled the feature-flag.

**Reply by BlackDex** [reply]
Have you tried to add an ssh-key? It might still be in there somewhere. That is probably causing an issue.

**Reply by twiesing** [reply]
Yes, that may be the problem. I tried to add it, but nothing happened when I clicked on ‘Select SSH key from clipboard’. It's possible that I clicked on Save afterwards anyway. The easiest way would probably be to search for it in the postgres and delete it?

**Reply by BlackDex** [reply]
Yes probably. The cipher type is `5`, if you delete that record it will probably be fixed.

**Reply by twiesing** [reply]
Deleted one entry from postgres db. Now the iOS app works :) Thank you @BlackDex and happy new year!

**Comment by Zen-Z** [comment]
I experienced the same issue. I also had stored an SSH key. After creating the SSH key on the desktop client, the iOS client started showing sync errors. The issue was resolved after deleting all the saved SSH keys.

**Environment**:  
Vaultwarden version: v1.32.7  

**Clients**:  
iOS  

**Client Version**:  
2024.12.0 (1730)
